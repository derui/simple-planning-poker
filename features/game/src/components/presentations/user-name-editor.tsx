import { Variant } from "@spp/shared-color-variant";
import { Icon } from "@spp/ui-icon";
import { Input } from "@spp/ui-input";
import * as styles from "./user-name-editor.css.js";

export interface Props {
  /**
   * the default value of this component
   */
  readonly defaultValue: string;

  /**
   * The callback on submit
   */
  readonly onSubmit?: (value: string) => void;

  /**
   * The callback on cancel
   */
  readonly onCancel?: () => void;
}

export const UserNameEditor = function UserNameEditor({ defaultValue, onSubmit, onCancel }: Props): JSX.Element {
  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const data = new FormData(event.currentTarget);
    const userName = data.get("userName") ?? "";

    onSubmit?.(userName.toString());
  };

  const handleCancel = (event: React.MouseEvent) => {
    event.preventDefault();
    onCancel?.();
  };

  return (
    <form className={styles.root} onSubmit={handleSubmit}>
      <span className={styles.decoration}>
        <Icon.Pencil size="m" variant={Variant.teal} />
      </span>
      <Input autoFocus name="userName" defaultValue={defaultValue} />
      <button type="submit" className={styles.submit} aria-label="Submit">
        <Icon.Check size="m" variant={Variant.emerald} />
      </button>
      <button type="button" className={styles.cancel} onClick={handleCancel} aria-label="Cancel">
        <Icon.X size="m" variant={Variant.cerise} />
      </button>
    </form>
  );
};
