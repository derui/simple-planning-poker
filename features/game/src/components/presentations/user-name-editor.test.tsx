import { cleanup, render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import Sinon from "sinon";
import { afterEach, expect, test } from "vitest";
import { UserNameEditor } from "./user-name-editor.js";

afterEach(cleanup);

test("should be able to render", async () => {
  const ret = render(<UserNameEditor defaultValue="foo" />);

  expect(ret.container).toMatchSnapshot();
});

test("should focus input first", () => {
  // Arrange
  render(<UserNameEditor defaultValue="foo" />);

  // Act
  const input = screen.getByRole("textbox");

  // Assert
  expect(input?.matches(":focus")).toBeTruthy();
});

test("should call submit callback after submitting", async () => {
  // Arrange
  const submit = Sinon.fake();
  render(<UserNameEditor defaultValue="foo" onSubmit={submit} />);

  // Act
  await userEvent.type(screen.getByRole("textbox"), "foobar");
  await userEvent.click(screen.getByRole("button", { name: "Submit" }));

  // Assert
  expect(submit.calledOnce).toBeTruthy();
  expect(submit.lastCall.args).toEqual(["foofoobar"]);
});

test("should cancel when cancel button is clicked", async () => {
  // Arrange
  const cancel = Sinon.fake();
  render(<UserNameEditor defaultValue="foo" onCancel={cancel} />);

  // Act
  await userEvent.click(screen.getByRole("button", { name: "Cancel" }));

  // Assert
  expect(cancel.called).toBeTruthy();
});
