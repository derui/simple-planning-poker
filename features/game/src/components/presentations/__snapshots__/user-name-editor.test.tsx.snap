// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to render 1`] = `
<div>
  <form
    class="user-name-editor_root__1isub0u0"
  >
    <div
      class="user-name-editor_row__1isub0u1"
    >
      <span
        class="user-name-editor_decoration__1isub0u3"
      >
        <span
          class="pencil_medium_teal__2ixuvsb style_iconMedium__p90ga82 style_iconBase__p90ga80"
        />
      </span>
      <input
        aria-invalid="false"
        class="style_input__xo256r0 style_transition_border__192bcwg42"
        name="userName"
        value="foo"
      />
    </div>
    <div
      class="user-name-editor_rowRight__1isub0u2"
    >
      <button
        aria-label="Submit"
        class="user-name-editor_submit__1isub0u4 style_enabledIconButtonVariants_emerald__92u52z10 style_enabledIconBase__92u52z5 style_iconBase__92u52z1 style_transition_all__192bcwg41"
        type="submit"
      >
        <span
          class="check_medium_emerald__1oavie2c style_iconMedium__p90ga82 style_iconBase__p90ga80"
        />
      </button>
      <button
        aria-label="Cancel"
        class="user-name-editor_cancel__1isub0u5 style_enabledIconButtonVariants_cerise__92u52z13 style_enabledIconBase__92u52z5 style_iconBase__92u52z1 style_transition_all__192bcwg41"
        type="button"
      >
        <span
          class="x_medium_cerise__15rhf1pf style_iconMedium__p90ga82 style_iconBase__p90ga80"
        />
      </button>
    </div>
  </form>
</div>
`;
